control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go

```diff
@@ -0,0 +1,351 @@
+func getOwnerPrefixFromPod(pod *corev1.Pod) string {
+	for _, ref := range pod.OwnerReferences {
+		if ref.Kind == "ReplicaSet" {
+			return ref.Name
+		}
+	}
+	return ""
+}
+
+// registerService creates a Consul service registration from the provided Kuberetes service and endpoint information.
+func (r *Controller) registerService(ctx context.Context, resourceClient pbresource.ResourceServiceClient, service corev1.Service, podPrefixes, podExactNames map[string]any) error {
+	serviceResource := r.getServiceResource(
+		&pbcatalog.Service{
+			Workloads:  getWorkloadSelector(podPrefixes, podExactNames),
+			Ports:      getServicePorts(service),
+			VirtualIps: r.getServiceVIPs(service),
+		},
+		service.Name, // Consul and Kubernetes service name will always match
+		r.getConsulNamespace(service.Namespace),
+		r.getConsulPartition(),
+		getServiceMeta(service),
+	)
+
+	r.Log.Info("registering service with Consul", getLogFieldsForResource(serviceResource.Id)...)
+	//TODO(zalimeni) Read from Consul or some other state to determine whether a write is needed.
+	// For now, we blindly write state on each reconcile to simplify implementation.
+	_, err := resourceClient.Write(ctx, &pbresource.WriteRequest{Resource: serviceResource})
+	if err != nil {
+		r.Log.Error(err, "failed to register service", getLogFieldsForResource(serviceResource.Id)...)
+		return err
+	}
+
+	return nil
+}
+
+// getServiceResource converts the given Consul service and metadata as a Consul resource API record.
+func (r *Controller) getServiceResource(svc *pbcatalog.Service, name, namespace, partition string, meta map[string]string) *pbresource.Resource {
+	return &pbresource.Resource{
+		Id:       getServiceID(name, namespace, partition),
+		Data:     common.ToProtoAny(svc),
+		Metadata: meta,
+	}
+}
+
+func getServiceID(name, namespace, partition string) *pbresource.ID {
+	return &pbresource.ID{
+		Name: name,
+		Type: &pbresource.Type{
+			Group:        "catalog",
+			GroupVersion: "v1alpha1",
+			Kind:         "Service",
+		},
+		Tenancy: &pbresource.Tenancy{
+			Partition: partition,
+			Namespace: namespace,
+		},
+	}
+}
+
+// getServicePorts converts Kubernetes Service ports data into Consul service ports.
+func getServicePorts(service corev1.Service) []*pbcatalog.ServicePort {
+	ports := make([]*pbcatalog.ServicePort, 0, len(service.Spec.Ports)+1)
+
+	for _, p := range service.Spec.Ports {
+		ports = append(ports, &pbcatalog.ServicePort{
+			VirtualPort: uint32(p.Port),
+			TargetPort:  p.TargetPort.String(),
+			Protocol:    common.GetPortProtocol(p.AppProtocol),
+		})
+	}
+
+	// Append Consul service mesh port in addition to discovered ports.
+	//TODO(zalimeni) Omit if zero mesh ports present in service endpoints? Or could there be some
+	// use of mesh-inject at the service level in the webhook as well?
+	ports = append(ports, &pbcatalog.ServicePort{
+		TargetPort: "mesh",
+		Protocol:   pbcatalog.Protocol_PROTOCOL_MESH,
+	})
```

Comment:
Assuming we support discovery sync (non-mesh use cases) in the future, I think we'd probably want to exclude this port addition - but it's probably something we can defer. Marked `TODO(zalimeni)` JIC we already know the answer before I ticket it as a follow-up.

Reply:
Marking this as a figure-out-later item for now